### DISCUSSION TITLE
CommandBar Extensions

### DISCUSSION THREAD

**Discussion by kazo0** [unoplatform/uno#6057] [Ideas]
The [CommandBar Extensions](https://platform.uno/docs/articles/controls/CommandBar.html#extensions) need to be cleaned up and some can be merged under better defined contracts like an AppBarButton. Below is a proposal of the new Attached Properties and some notes on how they would be implemented.

The naming of the `BackButtonTitle` will be changed to `CurrentPageTitle` to better reflect its actual behaviour. This property relates to the title of the back button on the _**next**_ page in the stack on iOS. So, setting `CurrentPageTitle` to "Test123" on Page A's `CommandBar` then forward-navigating to Page B from Page A will now set the back button's title to "Test123", indicating that you would be navigation _**back**_ to a page titled "Test123" (Page A).

## Option 1
We will consolidate most of the properties into the `NavigationCommand` Property and rename it to `NavigationButton`. The `CommandBarRenderer`s ([iOS](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Controls/CommandBarRenderer.iOS.cs)/[Android](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Controls/CommandBarRenderer.Android.cs)) will read the `Icon` and `Foreground` properties of the `AppBarButton` and map to the native properties of the native back button. 

|      Attached Property     	|      Type      	| Android 	| iOS 	|
|:--------------------------:	|:--------------:	|:-------:	|:---:	|
|      NavigationButton      	| `AppBarButton` 	|    X    	|  X  	|
| IsNavigationButtonAttached 	|     `bool`     	|    X    	|  X  	|
|      CurrentPageTitle      	|    `string`    	|         	|  X  	|
|          Subtitle          	|    `string`    	|    X    	|     	|

- The `NavigationButton`'s `Icon` and `Foreground` properties will be used to set the icon and colour of the native back button. The native back button will be completely replaced with this `AppBarButton` if `IsNavigationButtonAttached` is `true`.
- `IsNavigationButtonAttached`: Determines whether or not the AppBarButton is actually attached to the native CommandBar. This will replace the native back button and disabled the native back gesture.  
- `CurrentPageTitle` (iOS Only): Value to be used by the back button of the next page                                        
- `Subtitle` (Android Only): The subtitle is displayed below the native title. The colour of the subtitle can't be changed.  

## Option 2
We will keep the `NavigationCommand` behaviour the same, except for its name which will change to `NavigationButton`. Along with the `NavigationButton`, we will add a new `AppBarButton` called `BackButton` that will consolidation the current `BackButtonIcon` and `BackButtonForeground`

| Attached Property 	|      Type      	| Android 	| iOS 	|
|:-----------------:	|:--------------:	|:-------:	|:---:	|
|  NavigationButton 	| `AppBarButton` 	|    X    	|  X  	|
|     BackButton    	| `AppBarButton` 	|    X    	|  X  	|
|  CurrentPageTitle 	|    `string`    	|         	|  X  	|
|      Subtitle     	|    `string`    	|    X    	|     	|

- The `BackButton`'s `Icon` and `Foreground` properties will be used to set the icon and colour of the native back button 	
- `CurrentPageTitle` (iOS Only): Value to be used by the back button of the next page
- `Subtitle` (Android Only): The subtitle is displayed below the native title. The colour of the subtitle can't be changed.

**Comment by kazo0** [comment]
|      Attached Property     	|      Type      	| Android 	| iOS 	|
|:--------------------------:	|:--------------:	|:-------:	|:---:	|
|      NavigationButton      	| `AppBarButton` 	|    X    	|  X  	|
| NavigationButtonMode 	|     `NavigationButtonMode`     	|    X    	|  X  	|

`NavigationButtonMode` will be an enum with two values: `Native` and `Override`. `Native` will be the default mode, where we will not actually replace the native back button with the provided `NavigationButton`. `Override` will actually replace the native back button with the `NavigationButton`, in order to achieve things like custom logic on button pressed, burger menu, etc.

**Comment by robloo** [comment]
How does this relate with #6195 ?

Honestly, the WinUI CommandBar code is what is needed most. When this comes in, the existing (highly restrictive) CommandBar should likely be removed. The CommandBar implementation currently in Uno is likely best left to platform-specific code in my opinion. It functions like a MAUI wrapped control and isn't suited to the design objectives of Uno (pixel-perfect, WinUI compatible).

I expect Uno uses this implementation a lot internally so it could be pulled out into a separate control and left in. I just think it shouldn't be the default CommandBar implementation. It seems to create many more problems than it solves for those of us with existing code.
